`sync_read` and `bulk_read` return vec of results

omelia-iliffe opened the pull request:
The current behavior only captures one Err per call and returns that error even if other Dynamixels respond to the sync/bulk read.

The implementation that I have been using returns the much more verbose Vec of Results, one result for each Dynamixel responding, allowing for proper handling of each unresponsive Dynamixel in the sync/bulk.

Clippy is warning about very complex return types. Depending on what you think I could add some types to make the response types easier to read.
```rust
Result<Vec<Result<Response<Vec<u8>>, ReadError<T::Error>>>, crate::TransferError<T::Error>>
// becomes
type ResponseResult = Result<Response<Vec<u8>>, ReadError<T::Error>>
Result<Vec<ResponseResult>, crate::TransferError<T::Error>>

de-vri-es commented:
Hey!

I've been thinking a bit about this type of multi-motor response functions. The current approach is not ideal: the simple function hides some error details from you, or you have to use the callback version, which takes away flexibility in how you write your code.

But I also don't really like returning a `Result<Vec<Result<...>>`. That is also kinda confusing.

We could do something slightly more complicated to implement, but hopefully a lot easier to use: return an object to represent the whole transaction, with functions to get the next response. We can also implement `Iterator` for that type then:

```rust
let mut response = client.sync_read(.....)?;
while let Some((motor_id, response)) = response.next() {
    let response = response?;
    // Do stuff with the response.
}
```

What do you think? I'm thinking we can remove the other functions then: no callback or Vec version, only this.

Added advantage is that is maps better to async code (having an sync version is a future goal I want to tackle somehow, some time).

omelia-iliffe commented:
Yea cool, not something I had thought about but returning the vec is definitely  clunky and even more so if its a Vec of Results. 
So are you thinking that the iterator calls would actually perform the calls to `read_status_response` when called or would you do all the reading in one for and return the iterator afterwards.
If the prior, what happens if the response iterator is never used? The responses to the sync/bulk_read will still be in the buffer and will cause out of sync issues the next time a read is called.

I've also been thinking about an async version and would love to spend some time developing it. I have been given some time over the next few months to contribute to oss projects :)

Also I've implemented a [fast sync/bulk read](https://github.com/omelia-iliffe/dynamixel2-rs/tree/fast_read) but haven't tested it yet. This is my favourite crate to contribute to btw. I've learnt so much over the past year and can't really thank you enough!

de-vri-es commented:
> So are you thinking that the iterator calls would actually perform the calls to `read_status_response` when called or would you do all the reading in one for and return the iterator afterwards. If the prior, what happens if the response iterator is never used? The responses to the sync/bulk_read will still be in the buffer and will cause out of sync issues the next time a read is called. Is using `#[must_use]` enough?

I'm thinking the iterator will do the actual calls. It can finish reading any remaining responses in the `Drop` implementation.

This will not be so easy in a future async version (because we have no async drop), but that's a worry for then.

> I've also been thinking about an async version and would love to spend some time developing it. I have been given some time over the next few months to contribute to oss projects :)

Right, I would recommend not to start on this without quite some discussion first. I want to find a way to not duplicate the whole codebase. But I think all solutions to that also have serious drawbracks.

> Also I've implemented a [fast sync/bulk read](https://github.com/omelia-iliffe/dynamixel2-rs/tree/fast_read) but haven't tested it yet.

Sounds good! Always happy to get contributions. Just let me know when it's ready :)

> This is my favorite crate to contribute to btw. I've learnt so much over the past year and can't really thank you enough!

That's great to hear :D I'm always happy when other people get involved with projects I maintain. It's nice to know people are actually using it.

omelia-iliffe commented:
> Right, I would recommend not to start on this without quite some discussion first.  

Understood!
> I'm thinking the iterator will do the actual calls. It can finish reading any remaining responses in the Drop implementation.  

So I have started implementing a method that does this but can't return the type `Response<&[u8]>` as its borrowed from the `StatusPacket` owned by the current function.

EDIT: Nevermind I figure out a way, changing the From impl to use the data directly.
```rust
impl<'a> From<StatusPacket<'a>> for Response<&'a [u8]> {
	fn from(status_packet: StatusPacket<'a>) -> Self {
		Self {
			motor_id: status_packet.packet_id(),
			alert: status_packet.alert(),
			data: &status_packet.data[STATUS_HEADER_SIZE..],
		}
	}
}
```
I also discovered we can't impl `Iterator<Item=Response<&'a [u8]>` due to the lifetimes of 'a

Something related to this that I have been thinking about: 
There are *almost* duplicate functions for `u8`, `u16`, `u32` for every instruction. These types are also arbitrary as they don't necessary match the data type that is being read. ie I read a `u32` for position but then convert it to a `i32` as that the type the Dynamixel Ys use for position. 

I wonder if you would be open to the idea of letting the user define the type at call time and have traits two traits for Reading and Writing that data type. The trait impl would have a const `count` and a method like `to_bytes(self) -> &[u8]` and `try_from_bytes(&[u8]) -> Result<Self, _>`.
We could add impls for the primitive types and [u8; N] and users could extend it further.

One major draw back is that users have to specify a type when calling (`let pos: i32 = bus.read(1)`), however we could impl methods to match the existing `u8`, `u16, `u32` and more.

omelia-iliffe commented:
Hello, I've created a prototype for the SyncRead iterator and the `Read` / `Write` traits. Happy for some of this to get chucked out if its not the direction you think we should head. Just did it quickly so its all a bit rough at the moment.

I also moved the `StatusPacket`, `Response, and `InstructionPacket` into the `packet` mod to reduce the clutter in `bus.rs` and `device.rs`.

Somethings i'm not sure about:
 - I've added a `ParseError` as the `Read::try_from_bytes` method is fallible, but its currently opaque. Doesn't give the actual reason the parse failed,
 - To be able to implement iterator for the `SyncRead`  the `Data` type has to be known when creating it. The only practical way to specify this is with a turbofish: `bus.sync_read::<u8>`, we can keep the existing methods as well `bus.sync_read_u8`
 - You can still access the raw responses by calling `SyncRead::next_raw`  which will return `Response<&'a [u8]>`
 - I had initially returned `(u8, Response<_>)` with the `u8` being the motor_id as you typed out above, before realising the `motor_id` is a field of the `response`
 
Todo: 
- Decide on this approach
- Documentation
- Debug for `SyncRead`
- Migration of the rest of the methods (bulkread/write, syncwrite, regwrite)
- Do we want scan_cb to follow this iterator pattern as well?
- Name the Read/Write traits (maybe DataRead or Readable)

de-vri-es commented:
Hey! Thanks for starting on the implementation. Some of it conflicts with refactoring I started on ealier on a branch. I've merged those now and adjusted this branch to solve the merge conflicts.

I did also strip it down to only the bare minimum now: the `SyncRead` struct and a new `sync_read()` function signature.

I think some traits to allow generic decoding/encoding is also a good idea, but I want to re-integrate that in a separate commit.

I'm thinking to have two trait called:

```rust
trait Data<'a> {
  pub fn encode(buffer: &mut [u8]) -> Result<usize, ...>;
  pub fn decode(buffer: &'a [u8]) -> Result<(Self, usize), ...>;
}

trait FixedSizeData: Data {
  const ENCODED_SIZE: usize;
}
```

The idea is that we could also implement `Data` (but not `FixedSizeData`) for `&'a [u8]` and for `Vec<u8>`. Then we could have simple read/writes and bulk read/writes use `Data`, and have sync read/writes use `FixedSizeData`.

We could use some macro magic to implement the traits for all integer types.

omelia-iliffe commented:
Nice, looking much cleaner with the renaming branch merged!
I can re add the data trait with the signature you've suggested and integrate it into `SyncRead`.

omelia-iliffe commented:
Again something I've run into is that if we want to be able to impl `Iterator` for `SyncRead` the return type cannot borrow from `SyncRead` meaning returning `Response<&[u8]>` or `Response<Data<'_>>` isn't possible.
Does the `Data` trait need to have a lifetime? I can't see a use case it except for replacing `Response<&[u8]>`

de-vri-es commented:
It's possible to add the `impl` by requiring `T: Static` in addition to `T: Data<'a>`.

However, I'm not sure anymore if the lifetime is worth it. At-least for `sync_read()` we also need to know how much bytes to read, so we have to demand `FixedSizeData` or take a `count` parameter.

`FixedSizeData` can not be implemented for `&[u8]` or `Vec<u8>`, so we need a different API for reading one of those types anyway.

Probably that means there's no point in keeping the lifetime in the `Data` trait, so it just makes it unnecessarily complicated.

But I may want to think a little bit more about it.

omelia-iliffe commented:
> The idea is that we could also implement `Data` (but not `FixedSizeData`) for `&'a [u8]` and for `Vec<u8>`. Then we could have simple read/writes and bulk read/writes use `Data`, and have sync read/writes use `FixedSizeData`.

The simple read/writes also require the `count` parameter meaning that if you wanted to omit that from the fn signature you would need to use `FixedSizeData` anyway. If that's the case why not combine the two traits and remove the lifetime.

For the `&[u8]`, we can have methods like `read_response` and maybe `read_response_owned`. These options along with `read` which returns `Response<Data>` may cover all uses. What do you think?

de-vri-es commented:
> The simple read/writes also require the `count` parameter meaning that if you wanted to omit that from the fn signature you would need to use `FixedSizeData` anyway. If that's the case why not combine the two traits and remove the lifetime.
> 
> For the `&[u8]`, we can have methods like `read_response` and maybe `read_response_owned`. These options along with `read` which returns `Response<Data>` may cover all uses. What do you think?

I agree we'll need separate methods for `&[u8]`/`Vec<u8>`. However, with these traits we could still use a single `SyncRead` "iterator" struct for both. But I'm not sure if it's worth it to complicate the traits more.

de-vri-es commented:
> However, with a `Data<'a>` trait we could still use a single `SyncRead` "iterator" struct for both. But I'm not sure if it's worth it to complicate the traits more.

But on the other hand, it's nicer for `sync_read_bytes()` to use `From<&[u8]>` instead of `Data<'a>`.

de-vri-es commented:
Ok, I removed the lifetime from `Data` and added a `sync_read<T: Data>` and a `sync_read_bytes<T: From<&[u8]>`.

The one thing I don't like is the big amount of duplication between `SyncRead` and `SyncReadBytes`, but I do think that the public API is more important than a bit of code duplication.

omelia-iliffe commented:
This is looking great!

If we are happy with `SyncRead` and `SyncReadBytes` I can start looking at BulkRead.  
My thinking is that we only implement `BulkReadBytes` as having to use the same Data type for a `BulkRead` struct defeats one of its main purposes.

de-vri-es commented:
yeah, sounds good! I'm also quite happy how it looks. And it's equally useful on `no_std` :D

For bulk read, we can move the `T` parameter to the `next_response()` function. We can't implement `Iterator`, but I think that's okay.

One more multi response instruction we support is the broadcast ping (called `scan()` in the crate). We should also update that one to this pattern, I think :)

And we should fold all the single motor read and write functions into one read and one write function witht a generic parameter.

I can also do parts of that if you don't have the time, just let me know :)

omelia-iliffe commented on src/instructions/sync_read.rs:
```diff
@@ -1,192 +1,289 @@
-		}
-		Ok(())
+		let motor_id = self.pop_motor_id()?;
+		let response: Response<&[u8]> = match self.next_status_packet(motor_id) {
```
Could we add another private method that returns a result to reduce the Option<Result> methods?

omelia-iliffe replied:
Also this pattern of StatusPacket into Response<&[u8]> into Response<T> pops up a few times. 
I think added a helper fn would be beneficial.

de-vri-es replied:
> Could we add another private method that returns a result to reduce the Option methods?

Hmm, there would be only one place where that method would be called, right? Personally I don't see much value in splitting it out in that case, because if we split it we need to have good names and documentation for all of them (I also want to document internal function to help contributors (and myself) working on the code).

> Also this pattern of StatusPacket into Response<&[u8]> into Response pops up a few times.
> I think added a helper fn would be beneficial.

I was also thinking if we even still need `StatusPacket` and `Response<&[u8]>` as separate types. They're very similar. But I don't think we should merge them while this PR is open.

But I agree this ~could~ should be simplified. Maybe for now by directly implementing `TryFrom<StatusPacket>` for `Response<T: Data>`?

omelia-iliffe commented on src/instructions/write.rs:
```diff
@@ -11,46 +12,24 @@ where
-		Ok(read_response_if_not_broadcast(self, motor_id)?)
-	}
+			// TODO: handle this unwrap
+			data.encode(&mut buffer[2..]).unwrap();
```
As Data::Encode returns a result we may need to change the encode_parameters fn to also return a result.

omelia-iliffe replied:
Changing the signature has quite a big impact but I can't see another way? Any ideas?

omelia-iliffe replied:
It seems a shame not to handle to error as we handle every other `BufferToSmallError`. I changed the return type of `encode_parameters`, let me know if you want to keep it or I can remove that commit.

de-vri-es replied:
True, it's not that bad. Especially since they can only return a `BufferTooSmallError`!

omelia-iliffe commented:
I think the bulk read interface needs some more thought. 

> For bulk read, we can move the T parameter to the next_response() function. We can't implement Iterator, but I think that's okay.  

Having the T on the next_response function will mean that users have to correctly enter the byte count (as part of BulkReadData) and than ask for the correct data types in the correct order. I think that will be very error prone.   
```rust
let bulk_read = client
	.bulk_read(&[
		BulkReadData { motor_id: 1, address: 1, count: 2 },
		BulkReadData { motor_id: 2, address: 116, count: 4 },
	])
	.unwrap();
let first_res = bulk_read::<u16>.next();
let second_res = bulk_read::<u32>.next();
```
One might think that changing the type in the `next` call would correctly change the number of bytes read but it will instead cause an error or worse an incorrect value.  
It might make it more obvious if you have to specify a data type for each `BulkReadData`.  But it still has to be typed in two places in the correct order.
```rust
let bulk_read = client
	.bulk_read(&[
		BulkReadData::<u16> { motor_id: 1, address: 1 },
		BulkReadData::<u32> { motor_id: 2, address: 116 },
	])
	.unwrap();
let first_res = bulk_read::<u16>.next();
let second_res = bulk_read::<u32>.next();
```

de-vri-es commented:
Hmm, that's an interesting idea. But one complication is that a bulk read could also be used to read multiple adjacent registers per motor, each being potentially different data types.

And the return type of `next()` can't be different per invocation if we don't use a generic parameter. But if it is generic it can have a mismatch with the original request.

The only way to hide this kind of thing is to have something like a mutable `BulkRead` object that describes what to read, and will contain the received values once the function completes. This might be a bit complicated to implement, and to use though.

Maaaaybe we should skip `BulkRead` for this PR if its not going to be pretty much, so we can experiment with different interfaces without blocking this PR.

omelia-iliffe commented:
I added `BulkReadBytes` as this matches the replaces the existing methods without losing functionality. I think leaving it at that for now and experimenting with how `BulkRead` might work later is a good idea.

To make it easier to implement I changed the signature from `&'a [Read] where Read: AsRef<BulkReadData>` to bulk `&'a [BulkReadData]`. Working with multiple generics in `BulkReadBytes` was messy and I can't really see a reason for it needing `AsRef` as the `BulkReadData` struct is quite small. We could also derive `Copy` for it to make it easier to work with.

Nearly there! Just need to do some testing and check all the docs.

de-vri-es commented:
> I added `BulkReadBytes` as this matches the replaces the existing methods without losing functionality. I think leaving it at that for now and experimenting with how `BulkRead` might work later is a good idea.

Sounds good!

> To make it easier to implement I changed the signature from `&'a [Read] where Read: AsRef<BulkReadData>` to bulk `&'a [BulkReadData]`. Working with multiple generics in `BulkReadBytes` was messy and I can't really see a reason for it needing `AsRef` as the `BulkReadData` struct is quite small. We could also derive `Copy` for it to make it easier to work with.

Ah, yeah, good point. I think this is a left-over from when I was trying to accept any `Iterator<Item: AsRef<BulkReadData>>`. Problem is we iterate over it twice, so we would need to clone the iterator. I was a bit hesitant to do that, as you can then easily pass in `vec.into_iterator()`, which will end up unexpectedly cloning the whole vector.

The `AsRef` still means you can pass a slice of owned `BulkReadData` or a slice of references (or boxes, arcs, ...). But the latter is probably a pretty rare use-case.


> Nearly there! Just need to do some testing and check all the docs.

:partying_face:

omelia-iliffe commented:
Couple of things:
 - At the moment `reg_write` and `write` both take references to `T`. I assume this is desired as the methods don't need to own the data. It just makes it look a little strange when using primitives, ie `client.reg_write(1, 132, &2000u32)`
- I have removed `Client::write_bytes` and `Client::reg_write_bytes` its function is covered by write all ready.
- the `*_bytes` methods cannot return borrowed data. I get an error with trying `client.read_bytes::<&[u8]>(1, 132, 4)` as the lifetimes don't work. Is this the behavior you expected? It does mean there isn't a way to access borrow data from the buffers. 

Otherwise I think everything is looking good. I tested all the functions and they work correctly.

omelia-iliffe commented:
Added an improved integration_test system, that can switch between using Dynamixels or `MockDevice`s using the `integration_test` feature. IDs, baudrate, serialport path are configurable with env variables. 
I improved the `MockDevice` to be on par with a Dynamixel for most instructions. There is an issue preventing Sync/Bulk reads (and Ping but it doesn't cause the test to fail) which I will look at after this PR is merged.

The `integration_test` feature runs the tests with physical dynamixels, so I had to remove `--all-features` from the CI. This means the new features that get added won't be automatically tested unless the are default features :/

de-vri-es commented:
> Couple of things:
> 
>     * At the moment `reg_write` and `write` both take references to `T`. I assume this is desired as the methods don't need to own the data. It just makes it look a little strange when using primitives, ie `client.reg_write(1, 132, &2000u32)`

Yeah. Sadly we can't work around that by taking `impl AsRef<T>` either as that will break type deduction. We miiight be able to work around this by taking `T: Data` by value and implementing `Data` for references and values for integer types, but only for references for vectors and slices. This is certainly worth investigating, but not blocking for the PR (it might have more downsides than upsides also).

>     * I have removed `Client::write_bytes` and `Client::reg_write_bytes` its function is covered by write all ready.

Sounds good!

>     * the `*_bytes` methods cannot return borrowed data. I get an error with trying `client.read_bytes::<&[u8]>(1, 132, 4)` as the lifetimes don't work. Is this the behavior you expected? It does mean there isn't a way to access borrow data from the buffers.

I expected that to be possible. I'll try to figure out what is going wrong :)


> 
> Otherwise I think everything is looking good. I tested all the functions and they work correctly.

Nice! :D

de-vri-es commented on dynamixel2-cli/src/bin/dynamixel2/main.rs:
```diff
@@ -29,10 +29,13 @@ fn do_main(options: Options) -> Result<(), ()> {
+							Ok(r) => log_ping_response(&r, start.elapsed()),
+							Err(e) => log::error!("Error pinging motor: {}", e),
+						}
+					}
```
This looks so much nicer in to use! :partying_face:

de-vri-es commented:
> * I have removed `Client::write_bytes` and `Client::reg_write_bytes` its function is covered by write all ready.

Actually, I think this isn't true. `Vec<u8>` and `[u8]` don't implement `Data` because they are not fixed size data types.

de-vri-es commented:
Pushed some commits to:
* Check for motor errors when converting from StatusPacket to Response.
* Allow borrowing bytes from the read buffer (sadly requires separate functions using the Borrow trait).
* Re-add the `write_bytes()` function.
* Add read/write commands to the CLI (also to have rudimentary compile checks for reading/writing arbitrary number of bytes).

I didn't test this good enough yet though, it might be that the actual `read_bytes_borrow()` functions still have lifetime issues. I need to write more tests to verify these.

omelia-iliffe commented:
> I didn't test this good enough yet though, it might be that the actual read_bytes_borrow() functions still have lifetime issues. I need to write more tests to verify these.

Yes I couldn't get it to build. What are you imagining the `T` to be in these calls. I was able to build and test it if I replaced the T with `&[u8]` explicitly.

omelia-iliffe commented:
> I didn't test this good enough yet though, it might be that the actual read_bytes_borrow() functions still have lifetime issues. I need to write more tests to verify these.

Yes I couldn't get it to build. What are you imagining the `T` to be in these calls. I was able to build and test it if I replaced the T with `&[u8]` explicitly.

de-vri-es commented:
Merging this to make sure we don't get merge conflicts if other people start making PRs. If we need to change things we can do it with a follow-up PR :)
